mistralrs-core/src/ops.rs

```diff
@@ -0,0 +1,505 @@
+            candle_core::CpuStorage::F32(vs) => self.nonzero(vs, layout),
+            candle_core::CpuStorage::F64(vs) => self.nonzero(vs, layout),
+        };
+        let index_len = layout.dims().len();
+        let result_len = result.len() / index_len;
+        let result = CpuStorage::U32(result);
+        let shape = Shape::from_dims(&[result_len, index_len]);
+        Ok((result, shape))
+    }
+    #[cfg(feature = "cuda")]
+    fn cuda_fwd(
+        &self,
+        storage: &candle_core::CudaStorage,
+        layout: &Layout,
+    ) -> Result<(candle_core::CudaStorage, Shape)> {
+        if !layout.is_contiguous() {
+            return Err(candle_core::Error::RequiresContiguous { op: "nonzero" });
+        }
+        let dev = storage.device().clone();
+        let d_in = match storage.dtype() {
+            candle_core::DType::U8 => *storage.as_cuda_slice::<u8>()?.device_ptr(),
+            candle_core::DType::U32 => *storage.as_cuda_slice::<u32>()?.device_ptr(),
+            candle_core::DType::I64 => *storage.as_cuda_slice::<i64>()?.device_ptr(),
+            candle_core::DType::BF16 => *storage.as_cuda_slice::<bf16>()?.device_ptr(),
+            candle_core::DType::F16 => *storage.as_cuda_slice::<f16>()?.device_ptr(),
+            candle_core::DType::F32 => *storage.as_cuda_slice::<f32>()?.device_ptr(),
+            candle_core::DType::F64 => *storage.as_cuda_slice::<f64>()?.device_ptr(),
+        } as *const c_void;
+        let n = layout.shape().elem_count();
+        let num_nonzero = count_nonzero_cuda(storage.dtype(), d_in, u32::try_from(n)?);
+        let d_out = unsafe { dev.alloc::<u32>(num_nonzero as usize * layout.dims().len()) }
+            .map_err(|_| Error::Msg("Failed to allocate memory for nonzero result".to_string()))?;
+        let d_out_ptr = *d_out.device_ptr() as *mut c_void;
+        let dims = layout
+            .dims()
+            .iter()
+            .map(|&x| u32::try_from(x).unwrap())
+            .collect::<Vec<u32>>();
+        let d_dims = dev
+            .htod_copy(dims)
+            .map_err(|_| Error::Msg("Failed to copy dims to device".to_string()))?;
+        let d_dims_ptr = *d_dims.device_ptr() as *const c_void;
+        nonzero_cuda(
+            storage.dtype(),
+            d_in,
+            u32::try_from(n)?,
+            num_nonzero,
+            d_dims_ptr,
+            u32::try_from(layout.dims().len())?,
+            d_out_ptr,
+        );
+        let shape = Shape::from_dims(&[num_nonzero as usize, layout.dims().len()]);
+        let dst = candle_core::CudaStorage::wrap_cuda_slice(d_out, dev);
+        Ok((dst, shape))
+    }
+}
+
+pub trait NonZeroOp {
+    fn nonzero(&self) -> Result<Tensor>;
+}
+
+impl NonZeroOp for Tensor {
+    fn nonzero(&self) -> Result<Tensor> {
+        if !self.is_contiguous() {
+            return Err(candle_core::Error::RequiresContiguous { op: "nonzero" });
+        }
+        self.apply_op1_no_bwd(&NonZero {})
+    }
+}
+
+mod tests {
```

Comment:
I think there are missing tests for the bitwise operators here?

Reply:
add some tests. And I did find a bug😂